### ISSUE TITLE
Extrapolation problems 

### ISSUE THREAD

**Issue by moabe75** [CompPhysVienna/n2p2#62]
Hi Andreas,
I am trying to find a way to remove extrapolations from my NNP. It's been suggested that the extrapolating structures must be determined and added to the data set in order to remove the extrapolation warnings. I think I just need to write a script to extract those structures, compute the energy and forces and redo the training. I am sure your have faced this many times. I wonder if you have or know such a script. That would help me to save a lot of time. 

Many thanks.

**Comment by singraber** [comment]
Hi,

yes indeed that is a standard task when generating a NNP data set. I have written the tool `nnp-comp2` which can be used for this purpose. There is no documentation page yet, but here is a quick tutorial:

The tool `nnp-comp2` compares the data set predictions of two NNPs which were trained for the same system but differ in some settings, e.g. the number of neurons, the symmetry function setup or just the random seed for training. The idea behind this comparison is the following: if two NNPs agree on a predicted energy or force it is likely that the configuration has been seen in the training set. On the other hand, if the two NNPs do not agree, the configuration was not in the training data and should be included in an updated set. See [this paper, Fig. 3](https://doi.org/10.1103/PhysRevB.85.045439) for an illustration.

So the tool `nnp-comp2` does select new configurations based on three criteria:

* An extrapolation warning threshold
* An error difference threshold
* A force error threshold

The tool is executed in a two-step procedure: First, simply provide a data set `input.data` (coming from your MD trajectory) and two different NNPs `nnp-data-1/nnp-data-2` and run 

```
mpirun -np <even number of procs> nnp-comp2 compare
```
to create a comparison (see e.g. `diff.out`). In a second step use the comparison data to create new training data:

```
nnp-comp2 select <ew_threshold> <energy_threshold> <force_threshold> <list of elements present>
```
Have a look at the example in `examples/nnp-comp2`...

hope this helps for the moment...

**Comment by moabe75** [comment]
Many thanks Andreas. That seems to be an excellent tool. Let me ask a question: for the first step, I need to create a data set "input.data" from my MD trajectory. But, the problem for my case is that my data set are constructed only with 8 molecules and I am doing the MD simulations for 512 molecules. So, you see an inconsistency in the number of structures (in the data set and MD simulations) and it's not possible to add the extrapolating structures (with 512 molecules) into the data set.  
I am thinking of a solution. What if I do the MD simulations for 8 molecules to find those configurations were not in the training data and add them to the data set. Then try the updated NNP for 512 structures. Do you think this way would work?

**Comment by singraber** [comment]
Of course, if you cannot get reference data for the large system you will have to stick to the smaller simulation cell.. or at least choose some intermediate system size which you can still handle with the reference method.

**Comment by moabe75** [comment]
Good. There is another thing in this case that maybe you could help me. In Fig. 3 of that paper, two potential curves are exactly the same except the grey regions that show the extrapolating structures. But, in my case, two potentials are completely different (see the attached file). I only changed the random seed for the training. I am doing NpT simulations for the production run. 
Do I need to get them through another ensembles like NVE? To me, it shows the potential model is not reliable at all. 

![PE](https://user-images.githubusercontent.com/60357969/92497154-000f2580-f1c7-11ea-955f-a307100d2916.PNG)

**Comment by singraber** [comment]
It should definitely not look like that... did you maybe create two separate trajectories? The idea is not to use BOTH NNPs for creating trajectories but rather use only ONE! Then, recalculate the configurations along this trajectory with the other NNP. So in practice you create one trajectory, convert the configurations to the `input.data` format and hand it over to `nnp-comp2` which will recalculate and compare the predictions of both NNPs.

**Comment by moabe75** [comment]
I think you are right. I did one NpT simulation using the NNP1 and another simulation with the NNP2. So as you guessed, they are two separate trajectories. Now I understand how it works: comparing the predicted energy (and forces) by two different NNPs for the **_same_** configurations (or trajectory). 
Many thanks for your helpful explanation.

**Comment by moabe75** [comment]
Hi Andreas,
I've got some results, but I have a hard time to understand them. To me, something is not correct. Here what I did. I did NpT simulations for a box of 512 structures using NNP1 and selected 1000 configurations for the NNPs comparison. I trained another NNP (NNP2) by only changing the random seed. Here is the results:

NNP1:
Energy per atom in the Lammps simulations ≈ -220.5 eV (Energy per atom in the AIMD ≈ -220.47 eV)
Number of EWs  in the Lammps simulations = Zero

Energy per atom in the nnp comparison ≈ -220.27 eV
Number of EWs in the nnp comparison ≈ 3000-4000

NNP2:
Energy per atom in the nnp comparison = -220.24 eV
Number of EWs  in the nnp comparison = 3000-4000

1) The energies per atom for NNP1 are not the same as ones in the lammps simulations. I used exactly the same nnp for both.
2) No extrapolation warnings are observed during the lammps simulations. However, in the comparison there are a few thousands. 

Here is the NNPs comparison for ca. 400 configurations: 

![nnp_comp](https://user-images.githubusercontent.com/60357969/92937606-88850480-f419-11ea-8f44-7270cc410fd3.PNG)

It seems, NNP2 can be the same as NNP1 by a scaling factor. The shape of curves are almost identical. 

Here is the "diff.out" file:
[diff.txt](https://github.com/CompPhysVienna/n2p2/files/5209629/diff.txt)

I greatly appropriate your help. Many thanks.

**Comment by vsumaria** [comment]
What is the force & energy threshold being used for the selection in the nnp-comp2 tool? it is total energy difference or energy/atom & for forces, it is the mean force difference ?

**Comment by singraber** [comment]
For the energies it's a threshold regarding the energy difference per atom (here `s.energyRef` stores the difference in total energy of the two NNPs):

https://github.com/CompPhysVienna/n2p2/blob/9c350fbb14f45ed4aaccf0dd3a691ded2a3d0be0/src/application/nnp-comp2.cpp#L443-L448

For forces the threshold targets at the force difference of **individual** atoms in a structure, i.e. if **one or more** force differences exceed the threshold the whole structure is added. So it's not using an **averaged** force difference. Here again, `it->fRef` stores the force difference vector of an atom:

https://github.com/CompPhysVienna/n2p2/blob/9c350fbb14f45ed4aaccf0dd3a691ded2a3d0be0/src/application/nnp-comp2.cpp#L451-L465
